### PR TITLE
Handle token refund in MessageExecutor

### PR DESCRIPTION
**Pull request by Sajma27**
# Summary
A description of the changes proposed in the pull request.

Fixes #number

## Checklist
- [ ] Change is small and easy to review (please split big changes into multiple PRs)
- [ ] Change is consistent with architecture
- [ ] Change is consistent with test architecture
- [ ] Change is consistent with naming conventions
- [ ] New code is covered with tests
- [ ] Tests related to old code are updated
- [ ] Documentation is up to date with changes

### CODE REVIEW

File: universal-login-relayer/src/integration/ethereum/MessageExecutor.ts [hunk 1]
```diff
@@ -39,9 +43,15 @@ export class MessageExecutor implements IExecutor<SignedMessage> {
     }
   }
 
-  async execute(signedMessage: SignedMessage): Promise<providers.TransactionResponse> {
+  async execute(signedMessage: SignedMessage, tokenPriceInEth?: utils.BigNumberish): Promise<providers.TransactionResponse> {
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/src/integration/ethereum/MessageExecutor.ts:46]: `tokenPriceInEth` shouldn't be optional

File: universal-login-relayer/src/integration/ethereum/MessageExecutor.ts [hunk 2]
```diff
@@ -39,9 +43,15 @@ export class MessageExecutor implements IExecutor<SignedMessage> {
...
     await this.messageValidator.validate(signedMessage);
-    const transactionReq: providers.TransactionRequest = await this.walletContractService.messageToTransaction(signedMessage, '1');
+    ensure(tokenPriceInEth !== undefined, Error, 'Undefined tokenPriceInEth');
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/src/integration/ethereum/MessageExecutor.ts:48]: If we will require `tokenPriceInEth`, this check is redundant.

File: universal-login-relayer/src/integration/ethereum/MessageExecutor.ts [hunk 3]
```diff
@@ -39,9 +43,15 @@ export class MessageExecutor implements IExecutor<SignedMessage> {
...
+    await this.gasTokenValidator.validate({
+      gasPrice: signedMessage.gasPrice.toString(),
+      gasToken: signedMessage.gasToken,
+      tokenPriceInETH: tokenPriceInEth.toString(),
+    }, 0.3);
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/src/integration/ethereum/MessageExecutor.ts:53]: Add the last parameter gasOption = `cheap`

File: universal-login-relayer/src/integration/ethereum/MessageExecutor.ts [hunk 4]
```diff
@@ -24,8 +26,10 @@ export class MessageExecutor implements IExecutor<SignedMessage> {
 
   async handleExecute(messageHash: string) {
     try {
-      const signedMessage = await this.messageRepository.getMessage(messageHash);
-      const transactionResponse = await this.execute(signedMessage);
+      const messageItem = await this.messageRepository.get(messageHash);
+      const signedMessage = messageItem.message;
+      const tokenPriceInEth = messageItem.tokenPriceInEth;
+      const transactionResponse = await this.execute(signedMessage, tokenPriceInEth);
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/src/integration/ethereum/MessageExecutor.ts:32]: We can inline this 3 lines

File: universal-login-relayer/test/e2e/wallet.test.ts [hunk 5]
```diff
@@ -46,7 +46,7 @@ describe('E2E: Relayer - WalletContract routes', () => {
       operationType: OperationType.call,
       refundReceiver: deployer.address,
       gasToken,
-      gasPrice,
+      gasPrice: utils.parseUnits(gasPrice.toString(), 'gwei'),
```
Comment (xitronix) [inline universal-login-relayer/test/e2e/wallet.test.ts]: Revert this change. It should be fine without it

File: universal-login-relayer/test/e2e/wallet.test.ts [hunk 6]
```diff
@@ -64,7 +64,7 @@ describe('E2E: Relayer - WalletContract routes', () => {
     });
 
     it('free', async () => {
-      const msg = await getTransferMessage(0);
+      const msg = await getTransferMessage(utils.parseUnits('0', 'gwei'));
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/test/e2e/wallet.test.ts:67]:
```suggestion
      const msg = await getTransferMessage(0);
```

File: universal-login-relayer/test/integration/ethereum/MessageExecutor.test.ts [hunk 7]
```diff
@@ -1,37 +1,55 @@
...
 describe('INT: MessageExecutor', () => {
   let messageExecutor: MessageExecutor;
   let signedMessage: SignedMessage;
   let provider: providers.Provider;
   let wallet: Wallet;
   let walletContract: Contract;
+  let message: Message;
   const validator = {
-    validate: async () => {},
+    validate: () => {},
```
Comment (xitronix) [inline universal-login-relayer/test/integration/ethereum/MessageExecutor.test.ts:23]: We should keep type of fn
```suggestion
    validate: () => Promise.resolve() ,
```

File: universal-login-relayer/test/integration/ethereum/MessageExecutor.test.ts [hunk 8]
```diff
@@ -1,37 +1,55 @@
...
     const expectedBalance = (await provider.getBalance(signedMessage.to)).add(signedMessage.value);
-    const transactionResponse = await messageExecutor.execute(signedMessage);
+    const transactionResponse = await messageExecutor.execute(signedMessage, '1');
     await transactionResponse.wait();
     const balance = await provider.getBalance(signedMessage.to);
     expect(balance).to.eq(expectedBalance);
   });
+
+  it('should execute transaction for token and wait for it', async () => {
+    signedMessage = getTestSignedMessage({...message, gasPrice: TEST_GAS_PRICE_IN_TOKEN}, wallet.privateKey);
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/test/integration/ethereum/MessageExecutor.test.ts:43]: We should add here gasToken, right?

File: universal-login-relayer/test/integration/ethereum/MessageExecutor.test.ts [hunk 9]
```diff
@@ -1,37 +1,55 @@
...
+    const expectedBalance = (await provider.getBalance(signedMessage.to)).add(signedMessage.value);
+    const transactionResponse = await messageExecutor.execute(signedMessage, TEST_TOKEN_PRICE_IN_ETH);
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/test/integration/ethereum/MessageExecutor.test.ts:45]: We should also check, that refund was paid in token

File: universal-login-relayer/src/integration/ethereum/MessageExecutor.ts [hunk 10]
```diff
@@ -39,9 +41,15 @@ export class MessageExecutor implements IExecutor<SignedMessage> {
     }
   }
 
-  async execute(signedMessage: SignedMessage): Promise<providers.TransactionResponse> {
+  async execute(signedMessage: SignedMessage, tokenPriceInEth?: utils.BigNumberish): Promise<providers.TransactionResponse> {
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/src/integration/ethereum/MessageExecutor.ts:44]: let's pass here the whole messageItem

File: universal-login-relayer/test/core/services/execution/QueueService.test.ts [hunk 11]
```diff
@@ -45,14 +45,17 @@ describe('UNIT: Queue Service', () => {
   const minedTransactionHandler: any = {
     handle: onTransactionMined,
   };
+  const gasTokenValidator: any = {
+    validate: () => {},
```
Comment (xitronix) [inline universal-login-relayer/test/core/services/execution/QueueService.test.ts:49]: Validation is asynchronous
